scylla/src/transport/topology.rs

```diff
@@ -301,12 +371,274 @@ async fn query_keyspaces(conn: &Connection) -> Result<HashMap<String, Keyspace>,
+        let inner_type = &type_[langle_position + 1..type_.len() - 1];
+        let outer_type = &type_[..langle_position];
+        let type_ = match outer_type {
+            "map" | "tuple" => {
+                let types = inner_type.split(", ").collect::<Vec<_>>();
+                if outer_type == "map" {
+                    if types.len() != 2 {
+                        // FIXME: The correct error type is QueryError:ProtocolError but at the moment it accepts only &'static str
+                        return Err(QueryError::InvalidMessage(format!(
+                            "type {} is not implemented",
+                            type_
+                        )));
+                    }
+                    Map(
+                        Box::new(map_string_to_cql_type(conn, types[0])?),
+                        Box::new(map_string_to_cql_type(conn, types[1])?),
+                    )
+                } else {
+                    return Ok(CqlType::Tuple(
+                        types
+                            .iter()
+                            .map(|type_| map_string_to_cql_type(conn, type_))
+                            .collect::<Result<Vec<_>, _>>()?,
+                    ));
+                }
+            }
+            "set" => Set(Box::new(map_string_to_cql_type(conn, inner_type)?)),
+            "list" => List(Box::new(map_string_to_cql_type(conn, inner_type)?)),
+            _ => {
+                // FIXME: The correct error type is QueryError:ProtocolError but at the moment it accepts only &'static str
+                return Err(QueryError::InvalidMessage(format!(
+                    "type {} is not implemented",
+                    type_
+                )));
+            }
+        };
+        return Ok(CqlType::Collection { frozen, type_ });
+    }
+
+    Ok(match NativeType::from_str(type_) {
+        Ok(type_) => CqlType::Native(type_),
+        _ => CqlType::UserDefinedType {
+            frozen,
+            name: type_.to_string(),
+        },
+    })
+}
+
+async fn query_table_partitioner(
+    conn: &Connection,
+    keyspace_name: &str,
+    table_name: &str,
+) -> Result<Option<String>, QueryError> {
+    let mut partitioner_query =
+        Query::new("select partitioner from system_schema.scylla_tables where keyspace_name = ? and table_name = ?");
+    partitioner_query.set_page_size(1024);
+
+    let rows = conn
+        .query_all(&partitioner_query, (keyspace_name, table_name))
+        .await?
+        .rows
+        .ok_or(QueryError::ProtocolError(
+            "system_schema.scylla_tables query response was not Rows",
+        ))?;
+
+    Ok(rows
+        .into_typed::<(Option<String>,)>()
+        .next()
+        .ok_or(QueryError::DbError(
+            DbError::Invalid,
+            format!(
+                "no entry for table {}.{} in system_schema.scylla_tables",
+                keyspace_name, table_name
+            ),
+        ))?
```

Comment:
So it turns out that not all tables have a corresponding entry in system_schema.scylla_tables. Is it the expected behaviour?